### PR TITLE
Shared Memory Concurrent Map

### PR DESCRIPTION
**Pull request by dgarvit**
Shared Memory Concurrent Map (#16704)

Adds a fast, scalable, fine-grained concurrent map to the package
modules. Its API attempts to match the `Map` standard module, and
may be suitable as a replacement implementation or as reference
material for future efforts.

This effort has been long in the undertaking. A very special thanks
to @dgarvit for continuing to work on this PR and prepare it for
merging.

Thanks to @dgarvit, @LouisJenkinsCS!

Reviewed by @dlongnecke-cray.

TESTING

- [x] `library/packages/ConcurrentMap` on `linux64` and `darwin`
      (w/ `COMM=none` and `COMM=gasnet`)
- [x] Viewed docs locally

FUTURE WORK

- Explore using `ConcurrentMap` as a drop-in replacement for `Map`
- Get `ConcurrentMap` to work with `owned` and `shared`

---

(Original OP below this point.)

Follow up from #14409. Here is a summary of the changes made:
* Updated the code for Chapel 1.23
* Aligned the API structure to match that of Chapel standard map
* Added documentation
* Added some tests

### CODE REVIEW

**Comment by dlongnecke-cray** [comment]
Heads up, I'll start reviewing this sometime next week (probably Mon-Tues, don't want to keep you hanging @dgarvit). Thanks for all your hard work!

**Comment by bradcray** [comment]
@dlongnecke-cray: Pinging you on this as a reminder and assigned you as the owner based on your previous comment.

**Comment by dlongnecke-cray** [comment]
> @dlongnecke-cray: Pinging you on this as a reminder and assigned you as the owner based on your previous comment.

Half finished review, whoops 🖖 .

**Comment by bradcray** [comment]
@dlongnecke-cray : It's time for your every-three-month ping.  :)

**Comment by dlongnecke-cray** [comment]
This time, I come prepared!

**Comment by dlongnecke-cray** [comment]
Hey @dgarvit, it's been awhile. Are you up for doing some cleanup so that we can finally get this (beautiful!) thing merged? If you're busy and can't get around to it, I can probably take over from where you've left off (my goals being to get tests passing and merge things about as is).

Up to you.

**Comment by dgarvit** [comment]
Hey @dlongnecke-cray , glad to see we are finally going ahead with merging this! Sure, I am up for some cleanup. :)

**Comment by dlongnecke-cray** [comment]
Awesome, I'll check it out and add a review shortly. Thanks for being extra patient!

File: modules/packages/ConcurrentMap.chpl
Comment (dlongnecke-cray) [inline modules/packages/ConcurrentMap.chpl:2]: Let's go ahead and update this to the 2021 copyright.

**Review by dlongnecke-cray — CHANGES_REQUESTED** [review]
To start with, let's merge `master` into this branch and (my bad!) update that copyright one more time.

If not too much breaks then I'll check pull the changes and check out the docs.

**Comment by dlongnecke-cray** [comment]
Gentle poke @dgarvit.

**Comment by dgarvit** [comment]
Thanks for the reminder @dlongnecke-cray!

**Comment by dlongnecke-cray** [comment]
Hi @dgarvit, I'm a bit late to the party but am back at it again! I think this module is amazing, and that we definitely want to document it.

With that being said, let's go ahead and add `ConcurrentMap.chpl` to the list of `PACKAGES_TO_DOCUMENT` in `modules/Makefile`.

**Comment by dgarvit** [comment]
Added `ConcurrentMap.chpl` to the list. Running `make docs` is generating the following error:
```
/chapel/doc/rst/modules/packages/ConcurrentMap.rst:322:chpl:chplref reference target not found: ConcurrentMap(keyType, valType)
```
I am not sure what this means.

**Comment by dlongnecke-cray** [comment]
Yeah, so if you search for `ConcurrentMap(keyType, valType) in `ConcurrentMap.chpl`, you'll see that's listed as a `rtype` label for e.g. `extend`. That error just means a type matching `ConcurrentMap(type keyType, type valType)` couldn't be found. 

What we can do is just go ahead and remove that `rtype` label (this is what we do for other collections too, btw, so it's not a shortcoming). The reason why we remove it is because the type is generic.

**Comment by dlongnecke-cray** [comment]
@dgarvit If you'd like, we could skip working on the docs for this effort, and I could open a second PR where I make them public and call on you as a reviewer?

**Comment by dgarvit** [comment]
@dlongnecke-cray sorry for delay. Removing the `rtype` allowed the docs build to succeed. I pushed the change you requested.

> If you'd like, we could skip working on the docs for this effort, and I could open a second PR where I make them public and call on you as a reviewer?

I didn't quite understand, are you suggesting we work on the docs later (in a separate PR) and focus on rest of the PR first?

**Comment by dlongnecke-cray** [comment]
> @dlongnecke-cray sorry for delay. Removing the rtype allowed the docs build to succeed. I pushed the change you requested.

Great!!!

> I didn't quite understand, are you suggesting we work on the docs later (in a separate PR) and focus on rest of the PR first?

Yeah, since I figured you must be busy. I was offering for me to work on the docs in a second PR, and you would just review that PR to tell me what I got right or wrong.

**Comment by dgarvit** [comment]
Ah okay, great!

**Comment by dlongnecke-cray** [comment]
Looks like there are two failing tests:

```
[Error matching program output for library/packages/ConcurrentMap/testEquality]
[Error matching compiler output for library/packages/ConcurrentMap/testGetRemove]
```

I'll also start looking into them.

**Comment by dlongnecke-cray** [comment]
For `testEquality.chpl`, the error is:

```
> testEquality.chpl:20: error: halt reached - argument to ! is nil
```

From adding some writelns, I can see that every iteration over the maps is succeeding except the last, which is the `writeln(m1 != m2);`.

**Comment by dlongnecke-cray** [comment]
Throwing `--no-inline` and `--no-inline-iterators` helped get to the root of the issue. Looks like the error is when setting the linked list of deferred nodes on `ConcurrentMap.chpl:430`.

I tried adjusting that line to:

```chapel
if deferred != nil then deferred!.prev = deferredNode;
```

But after that change I'm now looping forever into that `else` block.

**Comment by bradcray** [comment]
@dlongnecke-cray : Given that that statement essentially seems to be part of a linked list insertion, I expect you'd need to guard a larger swath of code than just that one line with that conditional?

**Comment by dlongnecke-cray** [comment]
I'm not so sure, it looks like we're just building building up a doubly linked list of deferred buckets and that particular line sets the back pointer of the current head before replacing it (but only if the current head isn't `nil`).

I think the problem might be that we're not hitting a base case in that last branch, which is causing the while loop to spin forever. Trying to figure out what the base case should be now.

**Comment by dlongnecke-cray** [comment]
On line `484`, the logic that consumes the deferred list doesn't appear to update the head of the list, to my eyes. Shouldn't it update `deferred` while consuming it? It seems like it only updates a temporary reference `head`, and doesn't write back to `deferred` when it is done.

File: modules/packages/ConcurrentMap.chpl
Comment (dlongnecke-cray) [inline modules/packages/ConcurrentMap.chpl:430]: The test `testEquality` hits a `!` halt on this line. I think this line needs to check if deferred is `nil` before updating it. After I tried doing this I started looping forever on the `m1 != m2` line in the test.

File: modules/packages/ConcurrentMap.chpl
Comment (dlongnecke-cray) [inline modules/packages/ConcurrentMap.chpl:464]: After I adjusted line `430` I started spinning forever. You know this logic better than I do, but it seems to me we'd want to update `deferred` after consuming it (else we'll keep looping forever down below).

After I tried doing that I got stuck spinning on a CAS.

File: test/library/packages/ConcurrentMap/testAddSet.chpl
Comment (dlongnecke-cray) [inline test/library/packages/ConcurrentMap/testAddSet.chpl:3]: Let's adjust this to `new shared ConcurrentMap(int, int);`.

File: test/library/packages/ConcurrentMap/testClear.chpl
Comment (dlongnecke-cray) [inline test/library/packages/ConcurrentMap/testClear.chpl:4]: Let's adjust this to `new shared ConcurrentMap(int, int);`

File: test/library/packages/ConcurrentMap/testContains.chpl
Comment (dlongnecke-cray) [inline test/library/packages/ConcurrentMap/testContains.chpl:4]: Let's adjust this to be a `shared ConcurrentMap`.

File: test/library/packages/ConcurrentMap/testEquality.chpl
Comment (dlongnecke-cray) [inline test/library/packages/ConcurrentMap/testEquality.chpl:3]: Let's adjust both of these to be a `shared ConcurrentMap`.

File: test/library/packages/ConcurrentMap/testExtend.chpl
Comment (dlongnecke-cray) [inline test/library/packages/ConcurrentMap/testExtend.chpl:4]: Let's adjust both of these to be a `shared ConcurrentMap`.

File: test/library/packages/ConcurrentMap/testGetRemove.chpl
Comment (dlongnecke-cray) [inline test/library/packages/ConcurrentMap/testGetRemove.chpl:7]: Let's adjust this to say

```chapel
var m: ConcurrentMap(string, shared C);
```

**Review by dlongnecke-cray — CHANGES_REQUESTED** [review]
The tests need to prefix a management style before class types (let's used `shared`), because omitting one means "anymanaged" now. See comments. Doing this will fix the error with `testGetRemove.chpl`.

The bug with `testEquality` is more complicated. I've left some comments in the thread as I tried to fix it. Right now I'm stuck on a blocked CAS. You'll probably be able to figure it out faster than me, I could be missing something very obvious.

File: modules/packages/ConcurrentMap.chpl
Comment (dlongnecke-cray) [inline modules/packages/ConcurrentMap.chpl:915]: Is this method meant to be public? If not let's `pragma "no doc"` it.

**Review by dlongnecke-cray — CHANGES_REQUESTED** [review]
This next round of review is for the documentation of `ConcurrentMap`. In general I've asked you to do two things:

- If a class/module level variable/method is meant to be publicly visible, you should comment it (I believe you've already commented the stuff you've intended to be public)
- Otherwise, if it's just an implementation detail, you should attach the `pragma "no doc"` to it. This will make it not show up in generated documentation.

I've also asked you to include a module description at the top of the module. This could be something as simple as "implementation of a concurrent map", but if you're so inclined, feel free to describe it in more detail, perhaps even including:

- How parallel is the map? (e.g. lock free...)
- What kind of workloads does it excel with? Read heavy? Write heavy? Balanced?
- What research paper(s) or algorithm(s) inspired its design?
- Are there any constraints to using the map that users should be aware of? (e.g. epoch based memory)

I think these things would make an impressive data structure even more impressive, but they're not strictly required. Feel free to use as much or as little detail as you want.

**Comment by dlongnecke-cray** [comment]
Light poke @dgarvit :)

**Comment by dgarvit** [comment]
Thanks for the reminder @dlongnecke-cray ! I have started working on the next round of review. I'll try to wrap up and push the changes by the weekend :)

**Comment by dlongnecke-cray** [comment]
Awesome @dgarvit, so glad to hear it!

**Comment by dgarvit** [comment]
Hi @mppf ! Really sorry for the delay. I began with the changes, but got occupied with something else. Yes, I plan to finish the next steps to move this PR forward soon, maybe by this week.

**Comment by bradcray** [comment]
We're past feature freeze and I don't think this is ready to merge yet anyway, so I'm applying the post-release label here.

**Comment by dlongnecke-cray** [comment]
About time for me to poke you again @dgarvit! Do you think you'll have any time to work on this, soon? We're still very interested in having this as a package module.

File: test/library/packages/ConcurrentMap/testGetRemove.chpl
Comment (dlongnecke-cray) [inline test/library/packages/ConcurrentMap/testGetRemove.chpl:8]: This test is failing because `a` and `b` aren't shared. It turns out the map also has problems in `getAndRemove` because it tries to default-initialize non-nilable classes. Let's avoid this issue for now by making this a `ConcurrentMap(string, int)` instead.

**Review by dlongnecke-cray — CHANGES_REQUESTED** [review]
Hi @dgarvit, thanks for all your hard work. We're almost there, just a couple more changes!

**Comment by dgarvit** [comment]
@dlongnecke-cray I have pushed the changes requested!

**Review by dlongnecke-cray — APPROVED** [review]
Awesome work @dgarvit! I ran the tests and checked the docs and things are looking great. @bradcray Are you happy with the module docs? If so I am going to go ahead and merge this soon.

@dgarvit Would you be willing to write up a merge message?

**Comment by bradcray** [comment]
> . @bradcray Are you happy with the module docs? If so I am going to go ahead and merge this soon.

I don't feel the need to personally review all documentation (i.e., if you're happy with it, I am), but since you asked, I took a very quick look and caught these things:
* `upto` -> `up to`
* "200x faster than Chapel’s built-in map at 44 threads" — I wonder whether this is still true given that the built-in map has been improved significantly since then?  Most of those improvements have been in the scalar code, so it quite possibly is, but mostly...  I'm not sure that putting a number that's likely to become dated over time into the documentation is a good practice.  I'd probably just say something like "at the time of its development, it outperformed Chapel's built-in map when used in a multithreaded mode by orders of magnitude"?  Or just leave it to the paper to make that case.
* "Parallely" — isn't a word  ("Iterates over...  in parallel" maybe?)
* remove() has strange indentation on its description...  maybe a sphinx weirdness?

None of these are show-stoppers (it is a package module rather than a standard one), and we could also fix them after the merge.

Thanks for sticking with this @dgarvit!

**Comment by dlongnecke-cray** [comment]
@dgarvit Consider updating your OP with the message you want to be used as your merge message (if not I will create one as best I can)! Also consider (but not required) responding to Brad's feedback on docs as a last pass.

**Comment by dgarvit** [comment]
@dlongnecke-cray, sure, you can write the merge message. :)
I'll be pushing a change to update the docs soon!

**Comment by dlongnecke-cray** [comment]
Hey @dgarvit still wanna do that docs update? If not it's fine to do so as a followup. I'd like to merge this today!

**Comment by dgarvit** [comment]
Done!

**Comment by dlongnecke-cray** [comment]
🎉 🎉 🎉 🎉 

Thanks for all your help @dgarvit!

**Comment by dgarvit** [comment]
So happy to see this finally being merged! Thanks for all the support with this PR @dlongnecke-cray @bradcray !

**Pull request merged**
